File: backend/Testes/Comum/Builders/Dominio/Caracteristicas/MaoBuilder.cs
using Dominio.Sinais.Caracteristicas;

namespace Testes.Comum.Builders.Dominio.Caracteristicas
{
    public class MaoBuilder
    {
        private Dedo[] dedos;
        private double[] vetorNormalDaPalma;
        private double[] direcaoDaMao;
        private double[] posicaoDaPalma;

        public MaoBuilder()
        {
            dedos = DedoBuilder.DedosPadroes();
            vetorNormalDaPalma = new ArrayDeDouble().ComTamanho(3).Construir();
            direcaoDaMao = new ArrayDeDouble().ComTamanho(3).Construir();
            posicaoDaPalma = new ArrayDeDouble().ComTamanho(3).Construir();
        }

        public MaoBuilder ComDedos(Dedo[] dedos)
        {
            this.dedos = dedos;
            return this;
        }

        public MaoBuilder ComVetorNormalDaPalma(double[] vetorNormalDaPalma)
        {
            this.vetorNormalDaPalma = vetorNormalDaPalma;
            return this;
        }

        public MaoBuilder ComDirecaoDaMao(double[] direcaoDaMao)
        {
            this.direcaoDaMao = direcaoDaMao;
            return this;
        }

        public MaoBuilder ParaOIndice(int indice)
        {
            vetorNormalDaPalma = new double[] { indice, indice + 1, indice + 1 };
            direcaoDaMao = new double[] { indice + 1, indice, indice + 1 };
            posicaoDaPalma = new double[] { indice, indice, indice };
            var direcaoDosDedos = new double[] {indice + 1, indice, indice + 1};
            var posicaoDasPontasDosDedos = new double[] { indice + 2, indice + 2, indice + 2 };

            dedos = new[]
            {
                new DedoBuilder().DoTipo(TipoDeDedo.Dedao).ComPosicaoDaPonta(posicaoDasPontasDosDedos).ComDirecao(direcaoDosDedos).Construir(),
                new DedoBuilder().DoTipo(TipoDeDedo.Indicador).ComPosicaoDaPonta(posicaoDasPontasDosDedos).ComDirecao(direcaoDosDedos).Construir(),
                new DedoBuilder().DoTipo(TipoDeDedo.Meio).ComPosicaoDaPonta(posicaoDasPontasDosDedos).ComDirecao(direcaoDosDedos).Construir(),
                new DedoBuilder().DoTipo(TipoDeDedo.Anelar).ComPosicaoDaPonta(posicaoDasPontasDosDedos).ComDirecao(direcaoDosDedos).Construir(),
                new DedoBuilder().DoTipo(TipoDeDedo.Mindinho).ComPosicaoDaPonta(posicaoDasPontasDosDedos).ComDirecao(direcaoDosDedos).Construir()
            };

            return this;
        }

        public Mao Construir()
        {
            return new Mao
            {
                Dedos = dedos,
                VetorNormalDaPalma = vetorNormalDaPalma,
                PosicaoDaPalma = posicaoDaPalma,
                Direcao = direcaoDaMao
            };
        }
    }
}

File: frontend/js/tests/builders/MaoBuilder.js
var DedoBuilder = require('./DedoBuilder.js');

function MaoBuilder() {
    this._direcao = [0.0, 0.0, 0.0];
    this._vetorNormalDaPalma = [0.0, 0.0, 0.0];
    this._posicaoDaPalma = [0.0, 0.0, 0.0];
    this._velocidadeDaPalma = [0.0, 0.0, 0.0];
    this._raioDaEsfera = 0;
    this._pitch = function(){ return 0; };
    this._yaw = function(){ return 0; };
    this._roll = function(){ return 0; };
    
    this._dedos = [
        new DedoBuilder().construir(),
        new DedoBuilder().construir(),
        new DedoBuilder().construir(),
        new DedoBuilder().construir(),
        new DedoBuilder().construir()
    ];
}
MaoBuilder.prototype = {
    _direcao: undefined,
    _vetorNormalDaPalma: undefined,
    _dedos: undefined,
    _posicaoDaPalma: undefined,
    _velocidadeDaPalma: undefined,
    _raioDaEsfera: 0,
    _pitch: undefined,
    _yaw: undefined,
    _roll: undefined,

    comDirecao: function(direcao) {
        this._direcao = direcao;
        return this;
    },

    comVetorNormalDaPalma: function(vetorNormalDaPalma) {
        this._vetorNormalDaPalma = vetorNormalDaPalma;
        return this;
    },

    comDedos: function(dedos) {
        this._dedos = dedos;
        return this;
    },

    construir: function() {
        return {
            palmNormal: this._vetorNormalDaPalma,
            direction: this._direcao,
            palmVelocity: this._velocidadeDaPalma,
            palmPosition: this._posicaoDaPalma,
            stabilizedPalmPosition: this._posicaoDaPalma,
            sphereRadius: this._raioDaEsfera,
            pitch: this._pitch,
            roll: this._roll,
            yaw: this._yaw,
            fingers: this._dedos
        };
    }
};

module.exports = MaoBuilder;
